packages/web3-common/src/web3_base_provider.ts

```diff
@@ -0,0 +1,79 @@
+import {
+	JSONRPCPayload,
+	JSONRPCRequest,
+	JSONRPCResponseWithError,
+	JSONRPCResponseWithResult,
+	JSONRPCResult,
+} from './types';
+
+export interface ProviderMessage<T = JSONRPCResult> {
+	type: string;
+	data: T;
+}
+
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#connectivity
+export type Web3BaseProviderStatus = 'connecting' | 'connected' | 'disconnected';
+export type Web3BaseProviderCallback<T = JSONRPCResult> = (
+	error: Error | null,
+	result?: ProviderMessage<T>,
+) => void;
+
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#provider-errors
+export const JSONRPC_ERR_REJECTED_REQUEST = 4001;
+export const JSONRPC_ERR_UNAUTHORIZED = 4001;
+export const JSONRPC_ERR_UNSUPPORTED_METHOD = 4200;
+export const JSONRPC_ERR_DISCONNECTED = 4900;
+export const JSONRPC_ERR_CHAIN_DISCONNECTED = 4901;
+
+// Provider interface compatible with EIP-1193
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md
+export abstract class Web3BaseProvider {
+	// TODO: For legacy support, should be deprecated and removed in favor of `request`
+	public send<T = JSONRPCResult, T2 = unknown[]>(
+		payload: JSONRPCPayload<T2>,
+		callback: (
+			error?: JSONRPCResponseWithError<T>,
+			result?: JSONRPCResponseWithResult<T>,
+		) => void,
+	): void {
+		this.request<T, T2>({ method: payload.method, params: payload.params })
+			.then(d =>
+				callback(undefined, { result: d, id: payload.id ?? 0, jsonrpc: payload.jsonrpc }),
+			)
+			.catch(e => callback(e, undefined));
+	}
+
+	abstract getStatus(): Web3BaseProviderStatus;
+	abstract supportsSubscriptions(): boolean;
+
+	// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#request
+	abstract request<T = JSONRPCResult, T2 = unknown[]>(request: JSONRPCRequest<T2>): Promise<T>;
+
+	// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#events
+	abstract on<T = JSONRPCResult>(
+		type: 'message' | 'disconnect' | string,
+		callback: Web3BaseProviderCallback<T>,
+	): void;
+	abstract on(
+		type: 'connect' | 'chainChanged',
+		callback: Web3BaseProviderCallback<{
+			readonly [key: string]: unknown;
+			readonly chainId: string;
+		}>,
+	): void;
+	abstract on(
+		type: 'accountsChanged',
+		callback: Web3BaseProviderCallback<{
+			readonly [key: string]: unknown;
+			readonly accountsChanged: string[];
+		}>,
+	): void;
+	abstract removeListener(type: string, callback: Web3BaseProviderCallback): void;
+
+	abstract once?<T = JSONRPCResult>(type: string, callback: Web3BaseProviderCallback<T>): void;
+	abstract removeAllListeners?(type: string): void;
+	abstract connect(): void;
+	abstract disconnect(code: number, reason: string): void;
+	abstract reset(): void;
+	abstract reconnect(): void;
```

Comment:
I think these only make sense for web sockets. They aren't really useful for HTTP and are specified in EIP1193

Reply:
As I highlighted it couple of times earlier in discussion, EIP1193 standard or our provider abstract class both present a generic behaviour across all the providers regardless of any technology. 

As the EIP1193 mentions it as:

```
A JavaScript Ethereum Provider API for consistency across clients and applications.
```
So such protocol standards does not think of a particular technology implementation. 

e.g. If `HTTP` provider does not have ability to `reconnect`, it can still implement an empty function to make sure all the polymorphic interfaces using that abstract class works consistently.